src/MobileDevice/widget/MobileDevice.ts

```diff
@@ -0,0 +1,126 @@
+import * as dojoDeclare from "dojo/_base/declare";
+import * as domConstruct from "dojo/dom-construct";
+import * as WidgetBase from "mxui/widget/_WidgetBase";
+import * as dojoLang from "dojo/_base/lang";
+
+type OnGetDeviceInformationOptions = "doNothing" | "showPage" | "callMicroflow";
+
+class MobileDevice extends WidgetBase {
+    private deviceIDAttribute: string;
+    private deviceTypeAttribute: string;
+    private appVersionVersionAttribute: string;
+    private appVersionBuildAttribute: string;
+    private onGetDeviceInformationMicroflow: string;
+    private onGetDeviceInformationOption: OnGetDeviceInformationOptions;
+    private onGetDeviceInformationPage?: string;
+    private mxObject: mendix.lib.MxObject;
+
+    update(mxObject: mendix.lib.MxObject, callback: () => void) {
+        this.mxObject = mxObject;
+        const warnings = this.validateProps();
+        if (warnings) {
+            domConstruct.create("div", {
+                class: "alert alert-danger",
+                innerHTML: warnings
+            }, this.domNode);
+        } else {
+            this.setUpWidgetDom();
+            this.connect(document, "deviceReady", dojoLang.hitch(this, this.onDeviceReady));
+            if (!window.device) {
+                mxObject.set(this.deviceIDAttribute, "");
+                mxObject.set(this.deviceTypeAttribute, "Web");
+                this.save(mxObject);
+            }
+        }
+
+        callback();
+    }
+
+    uninitialize(): boolean {
+        window.removeEventListener("deviceReady", this.onDeviceReady);
+        return true;
+    }
+
+    private setUpWidgetDom() {
+        domConstruct.create("div", {
+            class: "mobile-device-widget",
+            id: "mobile-device"
+        }, document.body, "first");
+    }
+
+    private onDeviceReady() {
+        if (cordova.getAppVersion) {
```

Comment:
I changed the implementation to `if (cordova.hasOwnProperty("getAppVersion"))` . This checking is necessary to avoid the `mendix developer app` or any hybrid app that has not included the app version plugin from hanging/throwing an error.